### PR TITLE
Add `CartEventsContext` to dispatch events in the Cart block

### PR DESCRIPTION
**Pull request by opr**
This PR adds a new context called `CartEventsContext` it currently has two functions as its value: `onProceedToCheckout` and `dispatchOnProceedToCheckout`

It uses the same pattern as the `CheckoutEventsContext` however I opted to create this new context to keep the concerns separated.

In this PR, we also call `dispatchOnProceedToCheckout` when the proceed to checkout button is clicked, and if any observers throw an error, or return an object with a type of error causing the process to stop and the user will not proceed to checkout.

The `QuantitySelector` component subscribes to this event, and if the quantity does not satisfy the validation criteria then it focuses the input, returns the error object and stops the cart -> checkout flow.

Two tests are also included in this PR, one to check the event is dispatched and correctly handled by observers when pressing the button, and one to check unsubscribing from observers works correctly.

This PR is based on `remove/auto-quantity-update` (https://github.com/woocommerce/woocommerce-blocks/pull/7606).

#### Accessibility

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

https://www.loom.com/share/5445f3a871204aa7a6b89fbed6800889

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Install [WooCommerce Min/Max Quantities](https://woocommerce.com/products/minmax-quantities/)
2. Set up some items (please test with more than one) with minimum and maximum quantities. Add these items to your cart.
3. Go to the Cart block, edit the quantities so they violate the min/max rules you set. You should see an error.
4. Try to proceed to checkout. 
5. Ensure the **first** invalid input is focused.
6. Fix the first error and try again.
7. Ensure the next invalid input is focused.
8. Fix all errors and ensure you can proceed to checkout.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Prevent shoppers from proceeding to checkout if the cart has invalid items.

### CODE REVIEW

File: assets/js/blocks/cart/block.js [hunk 1]
```diff
@@ -87,8 +90,10 @@ const Block = ( { attributes, children, scrollToTop } ) => (
 		<StoreNoticesContainer context="wc/cart" />
 		<SlotFillProvider>
 			<CartProvider>
```
Comment (alexflorisca) [inline assets/js/blocks/cart/block.js:92]: I'm wondering if we should combine Cart and CartEvents providers? At the moment, the Cart provider just passed the children down and has a redirectUrl

**Review by alexflorisca — APPROVED** [review]
Nice work, I've tested and it works. Code looks good too, just one suggestion. I think you have quite a few TS errors though so worth fixing those before you merge. Not really sure if the bot displays the results after every push though...

**Review by mikejolley — APPROVED** [review]
👍🏻 I think it's ok to keep the cart provider separate to the cart events context provider. Maybe they will grow in scope over time.

File: assets/js/blocks/cart/block.js (hunk 1 above)
Comment (opr) [reply assets/js/blocks/cart/block.js:92]: Please see Mike's comment below

**Pull request merged**